site/content/blog/argocd-carvel-plugin.md

```diff
@@ -0,0 +1,159 @@
+      #! 3. Volume mount the custom binary to the bin directory
+      containers:
+        #@overlay/match by=overlay.subset({"name": "argocd-repo-server"}), expects=1
+        - name: argocd-repo-server
+          volumeMounts:
+            - mountPath: /usr/local/bin/ytt
+              name: custom-tools
+              subPath: ytt
+```
+
+### Make the carvel-ytt plugin available to Applications 
+Expose the plugin by patching the `argocd-cm` ConfigMap's `configManagementPlugins`. The command under `generate` is the command that will be run when using the plugin. 
+
+```yaml
+#! argocd-cm-overlay.yml
+#@ load("@ytt:overlay", "overlay")
+#@overlay/match by=overlay.subset({"kind":"ConfigMap", "metadata":{"name":"argocd-cm"}})
+---
+#@overlay/match missing_ok=True
+data:
+  #! TODO append instead of overwrite this
+  configManagementPlugins: |
+    - name: carvel-ytt
+      generate:                      # Command to generate manifests YAML
+        command: ["sh", "-c"]
+        args: ["ytt -f ."]
+```
+
+### Apply the changes to the cluster
+Since these overlays need to patch the Argo CD configuration, create the namespace and apply the overlays with the Argo CD installation manifests.
+
+```shell
+$ kubectl create namespace argocd
+```
+
+Now, lets install Argo CD Core. To keep setup simple use the core version that does not include the UI.
+
+Run this command to apply the ytt overlays to the Argo CD core manifests and then apply that to the cluster using kapp (or modify for kubectl, if you prefer). The `-y` continues deployment without confirmation. 
+
+```shell
+$ ytt -f argocd-cm-overlay.yml -f  repo-server-overlay.yml -f https://raw.githubusercontent.com/argoproj/argo-cd/v2.2.5/manifests/core-install.yaml  | kapp deploy --app argo --namespace argocd --file - -y
+```
+
+## Create and template an Application with the ytt plugin
+Now, create an Application resource that watches the git repo directory `config-step-2-template/`. This example contains simple ytt templates and data values for a Service and Deployment. If you want to see updates automatically deployed when you make changes, fork [this repo](https://github.com/vmware-tanzu/carvel-simple-app-on-kubernetes) and update the `repoURL` reference below. Add `spec.source.plugin` to tell Argo to use the ytt plugin.
+
+Additionally, create an `AppProject` resource to manually create the `default` project since there is a delay before the one that is automatically created is available. This shouldn't be necessary outside of this example.
+
+```yaml
+# simple-app-application.yml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: simple-app
+  namespace: argocd
+spec:
+  project: default
+  # this sync policy will deploy the app without confirmation when changes are detected
+  syncPolicy:
+    automated: {}
+  source:
+    repoURL: https://github.com/vmware-tanzu/carvel-simple-app-on-kubernetes.git
+    targetRevision: develop
+    path: config-step-2-template
```

Comment:
it would be great to show example how to pass configuration to ytt templates, or may be provide multiple directories (if this is even possible -- it may not be).

Reply:
> how to pass configuration to ytt templates

Do you mean how to pass flags like `--data-values-env`, for example?

In my research so far, it doesn't seem easily possible, but I can take another look.

Reply:
It may be possible using environment variables as suggested in this [Argocd issue](https://github.com/argoproj/argo-cd/issues/5510). I do not have a working example to share at this time. Could be an opportunity to mature this integration more at a later time.